modules/layers/src/icon-layer/icon-manager.js

```diff
@@ -0,0 +1,256 @@
+ * @returns {{mapping: {'/icon/1': {url, width, height, ...}},, canvasHeight: {Number}}}
+ */
+export function buildMapping({icons, buffer, maxCanvasWidth}) {
+  // x position till current column
+  let xOffset = 0;
+  // y position till current row
+  let yOffset = 0;
+  // height of current row
+  let rowHeight = 0;
+
+  let columns = [];
+  const mapping = {};
+
+  // Strategy to layout all the icons into a texture:
+  // traverse the icons sequentially, layout the icons from left to right, top to bottom
+  // when the sum of the icons width is equal or larger than maxCanvasWidth,
+  // move to next row starting from total height so far plus max height of the icons in previous row
+  // row width is equal to maxCanvasWidth
+  // row height is decided by the max height of the icons in that row
+  // mapping coordinates of each icon is its left-top position in the texture
+  for (let i = 0; i < icons.length; i++) {
+    const icon = icons[i];
+    if (!mapping[icon.url]) {
+      const {height, width} = icon;
+
+      // fill one row
+      if (xOffset + width + buffer > maxCanvasWidth) {
+        buildRowMapping(mapping, columns, yOffset);
+
+        xOffset = 0;
+        yOffset = rowHeight + yOffset + buffer;
+        rowHeight = 0;
+        columns = [];
+      }
+
+      columns.push({
+        icon,
+        xOffset
+      });
+
+      xOffset = xOffset + width + buffer;
+      rowHeight = Math.max(rowHeight, height);
+    }
+  }
+
+  if (columns.length > 0) {
+    buildRowMapping(mapping, columns, yOffset);
+  }
+
+  const canvasHeight = nextPowOfTwo(rowHeight + yOffset + buffer);
+
+  return {
+    mapping,
+    canvasHeight
+  };
+}
+
+// extract unique icons from data
+export function getIcons(data, getIcon) {
+  if (!data) {
+    return null;
+  }
+
+  const icons = {};
+  for (const point of data) {
+    const icon = getIcon(point);
+    if (icon && icon.url && !icons[icon.url]) {
+      icons[icon.url] = icon;
+    }
+  }
+  return icons;
+}
+
+export default class IconManager {
+  constructor(
+    gl,
+    {
+      data,
+      iconAtlas,
+      iconMapping,
+      getIcon = noop,
+      onUpdate = noop // notify IconLayer when icon texture update
+    }
+  ) {
+    this.gl = gl;
+    this.onUpdate = onUpdate;
+
+    this._mapping = {};
+    this._texture = null;
+
+    this._canvas = document.createElement('canvas');
+    this.updateState({data, iconAtlas, iconMapping, getIcon});
+  }
+
+  getTexture() {
+    return this._texture;
+  }
+
+  getIconMapping(dataPoint) {
+    const icon = this.getIcon(dataPoint);
+    const name = icon ? (typeof icon === 'object' ? icon.url : icon) : null;
+    return this._mapping[name] || {};
+  }
+
+  updateState({data, iconAtlas, iconMapping, getIcon}) {
+    if (getIcon) {
+      this.getIcon = getIcon;
+    }
+
+    if (iconMapping) {
+      this._mapping = iconMapping;
+    }
+
+    if (iconAtlas) {
+      this._updateIconAtlas(iconAtlas);
+    } else {
+      this._updateAutoPacking({
```

Comment:
`updateState` can be called without iconAtlas (e.g. when iconMapping changes) while still using prepacked icons.